### PR TITLE
daemon,overlord/assertstate: support streams of assertions with snap ack

### PR DESCRIPTION
**Pull request by pedronis**
This also introduce assertstate.Batch to handle such cases of out of order assertions to be added together to the system assertion database.

A later branch will switch firstboot code to use that as well.

### CODE REVIEW

File: overlord/assertstate/assertmgr.go [hunk 1]
```diff
@@ -101,12 +102,96 @@ func DB(s *state.State) asserts.RODatabase {
 	return cachedDB(s)
 }
 
-// Add the given assertion to the system assertiond database. Readding the current revision is a no-op.
+// Add adds the given assertion to the system assertion database.
```
Comment (chipaca) [inline overlord/assertstate/assertmgr.go]: Wait, `Add` adds? are you sure?

snark aside: i think you can (and should) drop the “adds”.

File: daemon/api.go [hunk 2]
```diff
@@ -1329,23 +1329,17 @@ func changeInterfaces(c *Command, r *http.Request, user *auth.UserState) Respons
 }
 
 func doAssert(c *Command, r *http.Request, user *auth.UserState) Response {
-	b, err := ioutil.ReadAll(r.Body)
+	batch := assertstate.NewBatch()
```
Comment (mvo5) [inline daemon/api.go]: Nice api!

File: overlord/assertstate/assertmgr.go [hunk 3]
```diff
@@ -101,12 +102,96 @@ func DB(s *state.State) asserts.RODatabase {
...
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		if err := b.Add(a); err != nil {
+			return nil, err
+		}
+	}
+	added := b.refs[start:]
```
Comment (mvo5) [inline overlord/assertstate/assertmgr.go]: Am I missing something or do we not test condition currently? start != 0?

Comment (pedronis) [inline overlord/assertstate/assertmgr.go]: you are right, added a test now, thank you

**Pull request merged**